### ISSUE TITLE
NCCL AllReduce deadlock on heterogeneous GPU clusters (GradScaler + DES-LOC sync desync)

### ISSUE THREAD

**Issue by zhao-zhaolang222** [dylanyunlon/Neuron_SP#2]
# Description

## Bug

DES-LOC training crashes with NCCL AllReduce timeout on heterogeneous GPU clusters when `torch.amp.GradScaler` skips `optimizer.step()` on a subset of ranks, causing `global_step` divergence across ranks and inconsistent sync decisions.

## Environment

- **Hardware**: 2×NVIDIA RTX A6000 (49GB) + 1×NVIDIA H100 NVL (96GB)
- **Software**: PyTorch 2.x, NCCL, torchrun
- **Config**: 700M model, Kx=32, Ku=96, Kv=192, batch=4×4
## Error

```
[DESLOC] Step 90/500 | Loss: 8.2812 | Time: 3531.7ms

[rank0]:[E424 04:11:42] Watchdog caught collective operation timeout:
WorkNCCL(SeqNum=71, OpType=ALLREDUCE, NumelIn=773615360,
Timeout(ms)=600000) ran for 600050 milliseconds before timing out.
```

Process killed after 10 minutes of AllReduce hang at step 90.

`DESLOCAdamW.sync_if_needed()` uses `self.global_step % Kx == 0` to decide whether to call `dist.all_reduce()`. `global_step` is incremented inside `optimizer.step()`. On heterogeneous GPUs:

1. A6000 (38.7 TFLOPS BF16) and H100 NVL (835 TFLOPS BF16) produce different gradient magnitudes for the same data
2. `torch.amp.GradScaler` detects NaN/Inf on one rank but not others → **skips `optimizer.step()` on that rank only**
3. Skipped rank's `global_step` falls behind by 1
4. At the next Kx boundary: some ranks compute `global_step % Kx == 0` → True (call AllReduce), others compute False (skip AllReduce)
5. NCCL AllReduce requires **all** ranks to participate → **deadlock**

This is specific to the combination of:
- Heterogeneous GPUs (different overflow probability per rank)
- `GradScaler` (skips step on overflow)
- DES-LOC gated sync (not every step calls AllReduce)

DDP is unaffected because it calls AllReduce every step regardless of optimizer state.

**Comment by dylanyunlon** [comment]
Bug no longer exists in current codebase:

1. **GradScaler removed** (L36: `GradScaler removed — BF16 does not need loss scaling`). BF16 training doesn't need loss scaling, eliminating the root cause of per-rank step skipping.

2. **Kx sync uses loop variable `step`**, not `self.global_step` (L1859: `_is_Kx_sync_pre = (step + 1) % self.desloc_Kx == 0`). The `for step in range(...)` loop variable is identical across all ranks, so sync decisions are always consistent.

3. **`global_step` set unconditionally** (L2294: `self.global_step = step + 1`) regardless of optimizer state, so it cannot diverge.

The original bug required the combination of: heterogeneous GPUs + GradScaler + gated sync. With GradScaler removed and sync decisions based on deterministic loop variables, the deadlock cannot occur.

Refs searched: PyTorch PR #116054 (consolidate FSDP ShardedGradScaler with amp.GradScaler)
